enumerable.rb

```diff
@@ -0,0 +1,159 @@
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    arr = [Hash, Range].member?(self.class) ? to_a.flatten : self
+    i = 0
+    while i < arr.length
+      yield(arr[i])
+      i += 1
+    end
+    arr
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    arr = [Hash, Range].member?(self.class) ? to_a.flatten : self
+    i = 0
+    while i < arr.length
+      yield(arr[i], i) if arr[i].is_a? Integer
+      i += 1
+    end
+    arr
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    arr = [Hash, Range].member?(self.class) ? to_a.flatten : self
+    results = []
+    arr.my_each { |x| results << x if yield(x) }
+    results
+  end
+
+  def my_all?(test = nil)
+    check = true
+    if block_given?
+      my_each { |x| check = false unless yield(x) }
+      check
+    elsif test.is_a? Class
+      my_each { |x| check = false if x.class != test }
+    elsif test.is_a? Regexp
+      my_each { |x| check = false unless test.match(x) }
+    else
+      t_ar = [nil, false]
+      my_each { |x| check = false if t_ar.include?(x) || x != test }
+    end
+    check
+  end
+
+  def my_any?(test = nil)
+    check = false
+    if block_given?
+      my_each { |x| check = true if yield(x) }
+    elsif test.is_a? Class
+      my_each { |x| check = true if x.class == test }
+    elsif test.is_a? Regexp
+      my_each { |x| check = true if test.match(x) }
+    else
+      test_array = [true, [], {}]
+      my_each { |x| check = true if test_array.include?(x) || include?(test) }
+    end
+    check
+  end
+
+  def my_none?(test = nil)
+    check = true
+    if block_given?
+      my_each { |x| check = false if yield(x) }
+    elsif test.is_a? Class
+      my_each { |x| check = false if x.class == test }
+    elsif test.is_a? Regexp
+      my_each { |x| check = false if test.match(x) }
+    else
+      my_each { |x| check = false if x == true || include?(test) }
+    end
+    check
+  end
+
+  def my_count(num = nil)
+    j = 0
+    if block_given?
+      my_each { |x| j += 1 if yield(x) }
+    elsif !num.nil?
+      my_each { |x| j += 1 if x == num }
+    else
+      return size
+    end
+    j
+  end
+
+  def my_map(proc = nil)
+    arr = is_a?(Range) ? to_a : self
+    results = []
+    if block_given?
+      arr.my_each { |x| results << yield(x) }
+    elsif proc
+      arr.my_each { |x| results << proc.call(x) }
+    else
+      return to_enum
+    end
+    results
+  end
```

Comment:
- [ ] The error reported by @maelfosso persist my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.